### ISSUE TITLE
collation: have some fun guys

### ISSUE THREAD

**Issue by Thorin-Oakenpants** [ghacksuserjs/TorZillaPrint#51]
Go [here](https://thorin-oakenpants.github.io/testing/collation.html)

@kkapsner @kolanich (and @ any watchers )

See if you can wrangle up some up some strings/chars. There's no entropy in collation in Firefox (or shouldn't be) if you're using `spoof_english` or a en-US build.

What I'm using this for to build a list of "items" that creates a different hash for all "locales" except `en` - if I can get one hash per locale that's even better: but probably not doable: e.g `nn` + `nb` = both Norwegian,

Then I can sort that array with `Intl.Collator` as `undefined` in FPing, and it should prove that extensions spoofing en-US is (except for en-GB and other en-* builds) utterly useless (although I have some other ways to leak this)

So far I'm using `44` strings/chars to detect `82` locales in `52` groups - as not `en`. Only 126 to go (we don't count `en`).

I'm struggling for some major european languages like spanish, french, german, portuguese

Feel free to have a play. I've included the missing collations, and if you look at the source, all the codes have their names provided (starts around line 50): e.g
```js
var list = [
		"ab", // abkhazian
		"ach",// acholi
		"aa", // afar
		"af", // afrikans
		"ak", // akan
		"ar", // arabic
		"am", // amharic
		"an", // arahonese
		"as", // assamese
		"ast",// asturian
```

so if you wanted to, you could research collation in arikans for me :) 👍 

PS: Am bitterly disappointed that `tlh` doesn't work. This is **klingon* .. but `nga` doesn't come after `no` .. and it should. I think Klingon is broken 👾

**Comment by Thorin-Oakenpants** [comment]
@kolanich .. what are you saying, that I need sha-256 to avoid collisions?

PS: 
- was:`44 items`, I'm up to `52` unique results for `82` locales .. `126` to go
- now:`50 items`, I'm up to `60` unique results for `85` locales .. `123` to go

I also cleaned up the output
- example outputs the locale/language name as well as the code
- no need to output the hash values
- missing and items use multiple lines

**Comment by KOLANICH** [comment]
>what are you saying, that I need sha-256 to avoid collisions?

Of course no, it's just embedded into the browser, so no need to  bring an impl of a hash function with oneself. sha1 is also embedded, but I guess it may be removed from crypto API in order to discourage its use.

**Comment by Thorin-Oakenpants** [comment]
oh. I already have the `sha1()` function on TZP in global.js - it's just I often build things as all-in-one html files before I work them upstream.

SO IDK .. if I should use the built-in one. I wonder if it's faster, or I could move everything to sha-256

**Comment by Thorin-Oakenpants** [comment]
I lost that link you gave me. I'm a tad confused - how do I call the function? 

some Links
- https://diafygi.github.io/webcrypto-examples/
- https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto

**Comment by kkapsner** [comment]
The basic German letter sorting is identical to the English one as far as I know. But we have 4 special characters: ä, ö, ü and ß
These have some special rules as they can be written as ae, oe, ue and ss and there should be differences between de-DE and de-AT (e.g. https://de.wikipedia.org/wiki/Alphabetische_Sortierung#Beispiel_f%C3%BCr_deutschsprachige_Sortierungen) - but in my quick test I didn't see any difference in the browser sorting.

PS: https://github.com/kkapsner/CanvasBlocker/blob/master/test/testAPI.js

**Comment by Thorin-Oakenpants** [comment]
german is starting to suck .. because it doesn't seem to be implemented

here's a snippet from [here](https://github.com/unicode-org/cldr/blob/master/common/collation/de.xml)
> - Thorn (U+00FE & U+00DE) is ordered as th, not as a first level letter
> - The letters wynn (U+01BF & U+01F7) and ezh (U+0292 & U+01B7) are ordered as y.
> - The Umlaut is the diacritic that must be taken into account before all others. The umlaut is treated as distinct from the trema(diaeresis) and can only occur in combination with the base letters a, o, and u.  The sequence of the other diacritics is compatible with the EOR delta.

But it doesn't work.

Meanwhile .. I'm at `170` locales differentiated from `en` .. up from 68 a few hours ago .. `38` to go

**Comment by kkapsner** [comment]
I think it's simply not that important as the standard sorting works good enough for German. The old germanic symbols (thorn, wynn and ezh) are not used in modern German and the Umlauts (actually the description from your link is wrong. The letters are Umlauts and not the diacritic...) are handled fine enough. And the handling of the ß is not consistent in daily use.

**Comment by Thorin-Oakenpants** [comment]
Dam my code is ugly as hell .. I was going to try to get the lines that have more than say eight locales, split. I can easily detect that but can't seem to loop an array inside an array or something.

Logic
- I calculate the hash for `en`
- I run through the `list` array and calculate a hash for each locale (I parse it for the short code)
- as I get a hash I compare `en` hash and do either
   - push it into the `results` array as `hash:locale+(long name)`
   - push it into the `missing` array as `locale`
- I sort the `results` array (i.e it is sorted by hash)
- I loop through `result` array --> `newArr` array which is `hash:locale1,locale2, etc`
- I loop through `newArr` array
   - for each item in the array I strip the first 41 chars to get the locale(s) string

It's during that last step, I know how many locales there are per hash (e.g. I can `.join` that string and get a length .. but anything after that ends up undefined: well, it works for the first two items and then fails

Got to be a better way to do this: although to be honest, once I get most of them into smaller lots, it will OK .. I just hate shit going off screen. I did try to make it word break all as a dirty hack. Oh well.

---

German. I dunno .. I'm just googling and githubbing and testing like mad. Tried the ss vs B symbol thing, etc. A lot of the stuff in there (and in articles) doesn't work. Like `cha`,`chz` for spanish = bollocks. Found an article on [Tibetan collation](https://collab.its.virginia.edu/wiki/tibetan-script/Sorting%20Tibetan.html) = all a load of shit :)

In the end I'm winging it from a 32K+ line txt file listing all unicode codes and their names

**Comment by Thorin-Oakenpants** [comment]
PS: you can run the example and copy/paste the chars in: it picks them up even if you can't display them :)

**Comment by Thorin-Oakenpants** [comment]
> - I loop through `result` array --> `newArr` array which is `hash:locale1,locale2, etc`
> - I loop through `newArr` array
>   - for each item in the array I strip the first 41 chars to get the locale(s) string

And now I see it, after typing it out. I can just output from result array: and as I hit locale8, locale15 add a line break to the output

Need MOAR coffee

**Comment by Thorin-Oakenpants** [comment]
OK, that's me for a few days - got other things to do

I added a display hash (a hash of the line number+ locales: ie what you see) so when you add/remove an item, you can see if anything actually changes

E.g. I just pasted in the example chars, and would remove something like `aa` and run, no change. Sweet. Got it down from 62 items to 49.

**Comment by Thorin-Oakenpants** [comment]
> Edit: I actually got down to SIX missing locales, but unfortunately that one chararcter I used also scooped up english :) I'll find the holy grail yet

That got me thinking: I explicitly check that the sort order is not the same as `en` .. so `en` **can't** be different... so I backtracked to that one character that caused it (turned out it was two characters), and added some debugging etc

```js
		//'\u0982',  //    BENGALI ANUSVARA
		//'\u0C02',  //     TELUGU ANUSVARA
```

These two, when added, detected 32 locales (including `en`: which it shouldn't) under one single hash. I output the `en` hash and sort array (and `en-US` for good measure), at the start, and during the loop to collect all hashes. The hash changed, and I could see one character getting put out of place.

Of those 32 locales, I moved `en`,`de`,`fr` to the very top of my list array (so they would be checked first), and lo and behold, they were no longer detected. So those characters are doing something weird and causing false positives. It's also telling that the 32 items were basically almost all consecutive entries from the list.

These were also added at the same time, and they too cause false positives

```js
		//'\u0902',  // DEVANAGARI ANUSVARA
		//'\u0A82',  //   GUJARATI ANUSVARA
		//'\u0C82',  //    KANNADA ANUSVARA
		//'\u0D02',  //  MALAYALAM ANUSVARA
		//'\u1036',  //    MYANMAR ANUSVARA
		//'\u0B02',  //      ORIYA ANUSVARA
		//'\u0D82',  //    SINHALA ANUSVARAYA
		//'\u0B82',  //      TAMIL ANUSVARA
```

I run `en` and `en-US` last so hopefully if anything is screwed up it will show - and I get a big red warning if they are detected. And I guess it just had to be too good to be true - finding a character that did so much. So I'll be mindful of any big changes per character.

tl;dr: **STAY TF away from anything ANUSVARA**

that's my story, guys :) Wasted half a day on that. Am back to 68 unique, 88 locales, 120 to go

**Comment by Thorin-Oakenpants** [comment]
closing this: in my local files (not the one online), I can detect
- FF: 80 hashes (97 locales) which don't match english
- chromium based browsers - it detects 38 hashes (45 locales)

Of the 59 collation locales not supported in chromium that FF does - three of them were added since FF60, add another three for good measure, and the TZP collation test (locally: will upload all my changes from the last month in one batch: so many improvements and new stuff, I've lost count) has been stripped down to six locales, six characters to speed up the test and is (should be: haven't tested safari)
- 100% reliable detection of FF
- as a bonus: shows version changes: e.g. FF70+, FF65-69, and FF64+

Its nice to add more entropy for show, but not neccesary since I have version detection anyway.

Finding more char strings to compare to detect other locales is becoming problematic, and I'm going to drop this avenue of FPing: Firefox users and TB on non en-US builds who use spoof_english are not affected. It was really only a way to demonstrate that using extensions to lie about the language is impossible to defend against (I have other methods besides this). Lying serves no purpose: if you successfully spoof anything, you're effectively just using en-US anyway: so why bother?

It's almost up there with UA spoofing as something incredibly stupid (except for compat)
